controllers/interfaces.go

```diff
@@ -0,0 +1,25 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import "context"
+
+// Reconciler provides operations for a reconciliation service
+type Reconciler interface {
```

Comment:
this looks very similar to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/cloud/interfaces.go#L28, I wonder if we should consider combining them.

Reply:
Should we just make `azuremachineService` also be of type `azure.Service`? As in it still does `reconcile` and `delete` like other cloud/services except that it is kind of a composite service which orchestrates other individual services?

Reply:
If I recall, @shysank made azuremachineService a azure.Service initially. I advised against that because I didn't want anyone to infer the azuremachineService was actually a cloud service, one of the services that lives `./cloud/services`.

I think the `Service` interface is really an abstract concept that overflows the bounds of `./cloud/services`. It is really a description of the reconcile behavior, reconcile to state or delete.

@CecileRobertMichon do you remember talking about how a reconciler could be a composition of reconiling services? I think this is that pattern is surfacing. `azuremachineService` is a reconciling service composed of a collection of reconciling services likely in some order of operations.

Perhaps, we just use `azure.Service` for now, and look to lift the interface from the `azure` package into a more generic package in a refactor.

Sorry if I've going into the bike shed on this one, but I just want to ensure we are making abstractions at the right levels.

Reply:
@devigned @CecileRobertMichon For now, I've made `azureMachineReconciler` to use `azure.service`. Filed an [issue](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1075) to move `azure.Service` outside of `cloud/`, PTAL.